contracts/NeapolitanMinionSummoner.sol

```diff
@@ -336,24 +373,28 @@ contract NeapolitanMinion is IERC721Receiver, IERC1155Receiver, IERC1271 {
         uint256 proposalId,
         address[] calldata actionTos,
         uint256[] calldata actionValues,
-        bytes[] calldata actionDatas) external memberOrModuleOnly returns (bool) {
+        bytes[] calldata actionDatas) external returns (bool) {
         Action memory action = actions[proposalId];
+        require(!action.executed, ERROR_EXECUTED);
+
+        if(action.memberOrModule) {
+            require(isMember(msg.sender) || msg.sender == module, ERROR_MEMBER_OR_MODULE_ONLY);
+        }
 
-        bool canExecute = isPassed(proposalId);
-        require(canExecute, ERROR_REQS_NOT_MET);
+        require(isPassed(proposalId), ERROR_REQS_NOT_MET);
         require(action.id != 0, ERROR_DELETED);
-        require(action.moloch == address(moloch), ERR_MOLOCH_CHANGED);
+        require(action.moloch == address(moloch), ERROR_MOLOCH_CHANGED);
 
         bytes32 id = hashOperation(actionTos, actionValues, actionDatas);
 
-        if(action.amount > 0) {
+        require(id == action.id, ERROR_NOT_VALID);
+        
+
+        if(action.amount > 0 && moloch.getUserTokenBalance(address(this), action.token) > 0) {
             // withdraw token tribute if any
             doWithdraw(action.token, moloch.getUserTokenBalance(address(this), action.token));
         }
         
-        require(id == action.id, ERROR_NOT_VALID);
-        require(!action.executed, ERROR_EXECUTED);
-        
         // execute calls
         actions[proposalId].executed = true;
```

Comment:
@dekanbro Is this just for re-entrancy protection? Because we delete the action down on line 406. Going to add a test to see what happens if you do an action which call execute action

Reply:
I think we should make this function nonreentrant anyway. Could be unexpected behavior if proposal 1 calls 'execute proposal' on proposal 0... will test

Reply:
On testing I realized it can't re-enter because it's self-referential and you can't encode the action to execute itself. but it could still call execute action on another proposal, probably expected behavior though

Reply:
i think we should move this above doWithdraw